src/java_bytecode/java_types.cpp

```diff
@@ -158,13 +167,54 @@ exprt java_bytecode_promotion(const exprt &expr)
     return typecast_exprt(expr, new_type);
 }
 
-typet java_type_from_string(const std::string &src)
+/// Transforms a string representation of a Java type into an internal type
+/// representation thereof.
+///
+/// Example use are object types like "Ljava/lang/Integer;", type variables like
+/// "TE;" which require a non-empty \p class_name or generic types like
+/// "Ljava/util/List<T>;" or "Ljava/util/List<Integer>;"
+///
+/// \param src: the string representation as used in the class file
+/// \param class_name: name of class to append to generic type variables
+/// \returns internal type representation for GOTO programs
+typet java_type_from_string(
+  const std::string &src,
+  const std::string &class_name)
 {
   if(src.empty())
     return nil_typet();
 
+  // a java type is encoded in different ways
+  //  - a method type is encoded as '(...)R' where the parenthesis include the
+  //    parameter types and R is the type of the return value
+  //  - atomic types are encoded as single characters
+  //  - array types are encoded as '[' followed by the type of the array
+  //    content
+  //  - object types are named with prefix 'L' and suffix ';', e.g.,
+  //    Ljava/lang/Object;
+  //  - type variables are similar to object types but have the prefix 'T'
   switch(src[0])
   {
+  case '<':
+    {
+      // The method signature can optionally have a collection of formal
+      // type parameters (e.g. on generic methods on non-generic classes
+      // or generic static methods). For now we skip over this part of the
+      // signature and continue parsing the rest of the signature as normal
```

Comment:
maybe give a concrete example here what would be skipped
could this '<' case and the '<' case inside 'L' - object name parsing be merged ?

Reply:
No - this should only occur for functions that look like:

```java
void <T> foo(T t)
```

etc (i.e. functions that are generic as opposed to part of a generic class). Perhaps a more robust way of dealing with this is to throw and exception and fall back to the descriptor in this case (since this is broadly what this is doing anyway)

Reply:
Proposed comment extension:
```C++
// So for example, the following java method:
// static void <T, U> foo(T t, U u, int x);
// Would have a signature that looks like:
// <T:Ljava/lang/Object;U:Ljava/lang/Object;>(TT;TU;I)V
// So we skip all inside the angle brackets and parse the rest of the string:
// (TT;TU;I)V
```

Reply:
Thank you Thomas, adding it in..